ESP32 with WiFi and threading builds but crashes at startup

### Summary

Testing with `examples/udp-echo`:

* `CONFIG_WIFI_NETWORK=... CONFIG_WIFI_PASSWORD=... laze build -b espressif-esp32-devkitc -s sw/threading run` builds and flashes, but then crashes with this output below
* Running without sw/threading works fine (and its output looks visually the same up to and including the EspWifiConfig line)

This topic was last updated in https://github.com/ariel-os/ariel-os/pull/777.

### Relevant log output

```shell
ets Jul 29 2019 12:21:46

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:2
load:0x3fff0030,len:7104
load:0x40078000,len:15576
load:0x40080400,len:4
ho 8 tail 4 room 4
load:0x40080404,len:3876
entry 0x4008064c
I (31) boot: ESP-IDF v5.1-beta1-378-gea5e0ff298-dirt 2nd stage bootloader
I (31) boot: compile time Jun  7 2023 07:48:23
I (33) boot: Multicore bootloader
I (37) boot: chip revision: v3.1
I (41) boot.esp32: SPI Speed      : 40MHz
I (46) boot.esp32: SPI Mode       : DIO
I (50) boot.esp32: SPI Flash Size : 4MB
I (55) boot: Enabling RNG early entropy source...
I (60) boot: Partition Table:
I (64) boot: ## Label            Usage          Type ST Offset   Length
I (71) boot:  0 nvs              WiFi data        01 02 00009000 00006000
I (79) boot:  1 phy_init         RF data          01 01 0000f000 00001000
I (86) boot:  2 factory          factory app      00 00 00010000 003f0000
I (94) boot: End of partition table
I (98) esp_image: segment 0: paddr=00010020 vaddr=00000001 size=000ach (   172) 
I (106) esp_image: segment 1: paddr=000100d4 vaddr=3ffb0000 size=02624h (  9764) load
I (118) esp_image: segment 2: paddr=00012700 vaddr=3ffda07c size=00214h (   532) load
I (123) esp_image: segment 3: paddr=0001291c vaddr=40080000 size=0b9b4h ( 47540) load
I (151) esp_image: segment 4: paddr=0001e2d8 vaddr=3f40e2d8 size=105ach ( 66988) map
I (176) esp_image: segment 5: paddr=0002e88c vaddr=4008b9b4 size=012f8h (  4856) load
I (178) esp_image: segment 6: paddr=0002fb8c vaddr=00000000 size=0048ch (  1164) 
I (182) esp_image: segment 7: paddr=00030020 vaddr=400d0020 size=5f004h (389124) map
I (337) boot: Loaded app from partition at offset 0x10000
I (337) boot: Disabling RNG early entropy source...
espressif-esp32-devkitc::init()
INFO esp-wifi configuration EspWifiConfig { rx_queue_size: 5, tx_queue_size: 3, static_rx_buf_num: 10, dynamic_rx_buf_num: 32, static_tx_buf_num: 0, dynamic_tx_buf_num: 32, ampdu_rx_enable: true, ampdu_tx_enable: true, amsdu_tx_enable: false, rx_ba_win: 6, max_burst_size: 1, country_code: "CN", country_code_operating_class: 0, mtu: 1492, tick_rate_hz: 100, listen_interval: 3, beacon_timeout: 6, ap_beacon_timeout: 300, failure_retry_cnt: 1, scan_method: 0 }
panic: panicked at /home/chrysn/.cargo/git/checkouts/esp-hal-70867c468ff64d88/aa8d199/xtensa-lx-rt/src/exception/context.rs:148:5:
Exception: LoadStoreError, Context { PC: 40080e33, PS: 00060815, A0: 800ebf2d, A1: 3ffb98d0, A2: 3f40f3c8, A3: 3ffd7918, A4: 00000001, A5: 3ffb1568, A6: 00000066, A7: 3ffb98d0, A8: 00000001, A9: 3ffb98b0, A10: 00060800, A11: 3ffd88e8, A12: 0000000a, A13: 00000000, A14: 3f413220, A15: 3ffb98b0, SAR: 00000020, EXCCAUSE: 00000003, EXCVADDR: 3f40f3c8, LBEG: 4000c2e0, LEND: 4000c2f6, LCOUNT: 00000000, THREADPTR: 00000000, SCOMPARE1: 00000100, BR: 00000000, ACCLO: 00000000, ACCHI: 00000000, M0: 00000000, M1: 00000000, M2: 00000000, M3: 00000000, F64R_LO_CPENABLE: 00000000, F64R_HI: 00000000, F64S: 00000000, FCR: 00000000, FSR: 00000000, F0: 00000000, F1: 00000000, F2: 00000000, F3: 00000000, F4: 00000000, F5: 00000000, F6: 00000000, F7: 00000000, F8: 00000000, F9: 00000000, F10: 00000000, F11: 00000000, F12: 00000000, F13: 00000000, F14: 00000000, F15: 00000000 }
```

### Hardware/ Setup

_No response_

### Possible Solution

_No response_

### Would you like to work on fixing this bug?

Yes

--- a/src/ariel-os-esp/src/preempt.rs
+++ b/src/ariel-os-esp/src/preempt.rs
@@ -3,14 +3,16 @@
 
 #![expect(unsafe_code)]
 
-use core::ffi::c_void;
+use core::{ffi::c_void, sync::atomic::AtomicUsize};
 
 use ariel_os_debug::log::{debug, trace};
 use ariel_os_threads::{THREAD_COUNT, create_raw, current_tid, yield_same};
 use esp_wifi::{TimeBase, preempt::Scheduler};
 use esp_wifi_sys::include::malloc;
 
-static THREAD_SEMAPHORES: [usize; THREAD_COUNT] = [0; THREAD_COUNT];
+/// The `AtomicUsize`s make sure this static is not promoted to read-only data.
+static THREAD_SEMAPHORES: [AtomicUsize; THREAD_COUNT] =
+    [const { AtomicUsize::new(0) }; THREAD_COUNT];
 
 struct ArielScheduler {}
 
@@ -52,17 +54,15 @@ impl Scheduler for ArielScheduler {
         let prio = ariel_os_embassy_common::executor_thread::PRIORITY;
         let core_affinity = None;
 
+        // SAFETY: *transmuting* between any two function pointers is fine.
+        let task =
+            unsafe { core::mem::transmute::<extern "C" fn(*mut c_void), extern "Rust" fn()>(task) };
+
         // SAFETY: Upholding `create_raw()` invariants: We know what we are doing.
-        let tid = unsafe {
-            create_raw(
-                task as usize,
-                param as usize,
-                stack_slice,
-                prio,
-                core_affinity,
-            )
-        };
-        usize::from(tid) as *const usize as *mut c_void
+        let tid =
+            unsafe { create_raw(task, Some(param as usize), stack_slice, prio, core_affinity) };
+
+        usize::from(tid) as *mut c_void
     }
 
     fn schedule_task_deletion(&self, _task_handle: *mut c_void) {
@@ -80,7 +80,8 @@ impl Scheduler for ArielScheduler {
         // NOTE(no-panic): this is always called from within a thread, so the `unwrap()` doesn't
         // panic.
         let tid = usize::from(current_tid().unwrap());
-        &THREAD_SEMAPHORES[tid] as *const usize as *mut c_void
+
+        THREAD_SEMAPHORES[tid].as_ptr() as *mut c_void
     }
 }
 

